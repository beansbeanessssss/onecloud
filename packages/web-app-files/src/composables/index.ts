export * from './fileListHeaderPosition'
export * from './pagination'
export * from './resourcesViewDefaults'
export * from './router'
export * from './selection'
export * from './sort'
export * from './viewMode'
