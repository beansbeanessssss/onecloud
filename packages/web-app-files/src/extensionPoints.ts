import { ExtensionPoint } from '@ownclouders/web-pkg'
import { computed } from 'vue'

export const uploadMenuExtensionPoint: ExtensionPoint = {
  id: 'app.files.upload-menu',
  extensionType: 'action',
  multiple: true
}
export const quickActionsExtensionPoint: ExtensionPoint = {
  id: 'app.files.quick-actions',
  extensionType: 'action',
  multiple: true
}
export const batchActionsExtensionPoint: ExtensionPoint = {
  id: 'app.files.batch-actions',
  extensionType: 'action',
  multiple: true
}
export const contextActionsExtensionPoint: ExtensionPoint = {
  id: 'app.files.context-actions',
  extensionType: 'action',
  multiple: true
}
export const fileSideBarExtensionPoint: ExtensionPoint = {
  id: 'app.files.sidebar',
  extensionType: 'sidebarPanel',
  multiple: true
}
export const folderViewsFolderExtensionPoint: ExtensionPoint = {
  id: 'app.files.folder-views.folder',
  extensionType: 'folderView'
}
export const folderViewsFavoritesExtensionPoint: ExtensionPoint = {
  id: 'app.files.folder-views.favorites',
  extensionType: 'folderView'
}
export const folderViewsProjectSpacesExtensionPoint: ExtensionPoint = {
  id: 'app.files.folder-views.project-spaces',
  extensionType: 'folderView'
}

export const extensionPoints = () => {
  return computed<ExtensionPoint[]>(() => {
    return [
      uploadMenuExtensionPoint,
      quickActionsExtensionPoint,
      batchActionsExtensionPoint,
      contextActionsExtensionPoint,
      fileSideBarExtensionPoint,
      folderViewsFolderExtensionPoint,
      folderViewsFavoritesExtensionPoint,
      folderViewsProjectSpacesExtensionPoint
    ]
  })
}
