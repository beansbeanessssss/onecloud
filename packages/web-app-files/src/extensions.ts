import {
  ApplicationSetupOptions,
  Extension,
  useStore,
  useRouter,
  useSearch,
  useFileActionsShowShares,
  useFileActionsCreateQuickLink
} from '@ownclouders/web-pkg'
import { computed, unref } from 'vue'
import { SDKSearch } from './search'

export const extensions = ({ applicationConfig }: ApplicationSetupOptions) => {
  const store = useStore()
  const router = useRouter()
  const { search: searchFunction } = useSearch()

  const { actions: showSharesActions } = useFileActionsShowShares()
  const { actions: quickLinkActions } = useFileActionsCreateQuickLink()

  return computed(
    () =>
      [
        {
          id: 'com.github.owncloud.web.files.search',
          type: 'search',
          searchProvider: new SDKSearch(store, router, searchFunction)
        },
        {
          id: 'com.github.owncloud.web.files.quick-action.collaborator',
          scopes: ['files', 'files.quick-action'],
          type: 'action',
          action: unref(showSharesActions)[0]
        },
        {
          id: 'com.github.owncloud.web.files.quick-action.quicklink',
          scopes: ['files', 'files.quick-action'],
          type: 'action',
          action: unref(quickLinkActions)[0]
        }
      ] satisfies Extension[]
  )
}
