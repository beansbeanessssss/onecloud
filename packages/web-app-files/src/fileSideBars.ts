import FileDetails from './components/SideBar/Details/FileDetails.vue'
import FileDetailsMultiple from './components/SideBar/Details/FileDetailsMultiple.vue'
import FileActions from './components/SideBar/Actions/FileActions.vue'
import FileVersions from './components/SideBar/Versions/FileVersions.vue'
import SharesPanel from './components/SideBar/Shares/SharesPanel.vue'
import NoSelection from './components/SideBar/NoSelection.vue'
import SpaceActions from './components/SideBar/Actions/SpaceActions.vue'
import {
  SpaceDetails,
  SpaceDetailsMultiple,
  SpaceNoSelection,
  isLocationTrashActive,
  isLocationPublicActive,
  isLocationSpacesActive,
  isLocationSharesActive,
  useCapabilityFilesSharing,
  useCapabilityFilesSharingApiEnabled,
  useCapabilityFilesSharingPublicEnabled,
  useRouter,
  useStore,
  SidebarPanelExtension,
  useIsFilesAppActive
} from '@ownclouders/web-pkg'
import {
  isProjectSpaceResource,
  spaceRoleEditor,
  spaceRoleManager
} from '@ownclouders/web-client/src/helpers'
import { Resource } from '@ownclouders/web-client'
import { useGettext } from 'vue3-gettext'
import { computed, unref } from 'vue'

export const sideBarPanels = () => {
  const router = useRouter()
  const store = useStore()
  const user = computed(() => store.getters.user)
  const { $gettext } = useGettext()
  const isSharingEnabled = useCapabilityFilesSharing(store)
  const isSharingApiEnabled = useCapabilityFilesSharingApiEnabled(store)
  const arePublicLinksEnabled = useCapabilityFilesSharingPublicEnabled(store)
  const isFilesAppActive = useIsFilesAppActive()

  return computed(
    () =>
      [
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.no-selection',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'no-selection',
            icon: 'questionnaire-line',
            title: () => $gettext('Details'),
            component: NoSelection,
            isRoot: () => true,
            isEnabled: ({ parent, items }) => {
              if (isLocationSpacesActive(router, 'files-spaces-projects')) {
                // project spaces overview has its own "no selection" panel
                return false
              }
              if (items?.length > 0) {
                return false
              }
              // empty selection in a project space root shows a "details" panel for the space itself
              return !isProjectSpaceResource(parent)
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.details-single-selection',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'details',
            icon: 'questionnaire-line',
            title: () => $gettext('Details'),
            component: FileDetails,
            componentAttrs: () => ({
              previewEnabled: unref(isFilesAppActive)
            }),
            isRoot: () => !isLocationTrashActive(router, 'files-trash-generic'),
            isEnabled: ({ items }) => {
              if (isLocationTrashActive(router, 'files-trash-generic')) {
                // details panel is not available in trash
                return false
              }
              if (items?.length !== 1) {
                return false
              }
              // project spaces have their own "details" panel
              return !isProjectSpaceResource(items[0])
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.details-multi-selection',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'details-multiple',
            icon: 'questionnaire-line',
            title: () => $gettext('Details'),
            component: FileDetailsMultiple,
            componentAttrs: () => ({
              get showSpaceCount() {
                return (
                  !isLocationSpacesActive(router, 'files-spaces-generic') &&
                  !isLocationSharesActive(router, 'files-shares-with-me') &&
                  !isLocationTrashActive(router, 'files-trash-generic')
                )
              }
            }),
            isRoot: () => true,
            isEnabled: ({ items }) => {
              if (isLocationSpacesActive(router, 'files-spaces-projects')) {
                // project spaces overview has its own "no selection" panel
                return false
              }
              return items?.length > 1
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.actions',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'actions',
            icon: 'slideshow-3',
            title: () => $gettext('Actions'),
            component: FileActions,
            isRoot: () => isLocationTrashActive(router, 'files-trash-generic'),
            isEnabled: ({ items }) => {
              if (items?.length !== 1) {
                return false
              }
              // project spaces have their own "actions" panel
              return !isProjectSpaceResource(items[0])
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.sharing',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'sharing',
            icon: 'user-add',
            iconFillType: 'line',
            title: () => $gettext('Shares'),
            component: SharesPanel,
            componentAttrs: () => ({
              showSpaceMembers: false,
              get showLinks() {
                if (unref(isSharingEnabled)) {
                  return unref(arePublicLinksEnabled)
                }
                return false
              }
            }),
            isEnabled: ({ items }) => {
              if (items?.length !== 1) {
                return false
              }
              if (isProjectSpaceResource(items[0])) {
                // project space roots have their own "sharing" panel (= space members)
                return false
              }
              if (
                isLocationTrashActive(router, 'files-trash-generic') ||
                isLocationPublicActive(router, 'files-public-link')
              ) {
                return false
              }
              if (unref(isSharingEnabled)) {
                return unref(isSharingApiEnabled)
              }
              return false
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.versions',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'versions',
            icon: 'git-branch',
            title: () => $gettext('Versions'),
            component: FileVersions,
            isEnabled: ({ items }) => {
              if (items?.length !== 1) {
                return false
              }
              if (isProjectSpaceResource(items[0])) {
                // project space roots don't support versions
                return false
              }
              if (
                isLocationTrashActive(router, 'files-trash-generic') ||
                isLocationPublicActive(router, 'files-public-link')
              ) {
                return false
              }
              return items[0].type !== 'folder'
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.projects.no-selection',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'no-selection',
            icon: 'questionnaire-line',
            title: () => $gettext('Details'),
            component: SpaceNoSelection,
            isRoot: () => true,
            isEnabled: ({ items }) => {
              if (!isLocationSpacesActive(router, 'files-spaces-projects')) {
                // only for project spaces overview
                return false
              }
              return items?.length === 0
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.projects.details-single-selection',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'details-space',
            icon: 'questionnaire-line',
            title: () => $gettext('Details'),
            component: SpaceDetails,
            isRoot: () => true,
            isEnabled: ({ items }) => {
              return items?.length === 1 && isProjectSpaceResource(items[0])
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.projects.details-multi-selection',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'details-space-multiple',
            icon: 'questionnaire-line',
            title: () => $gettext('Details'),
            component: SpaceDetailsMultiple,
            componentAttrs: ({ items }) => ({
              selectedSpaces: items
            }),
            isRoot: () => true,
            isEnabled: ({ items }) => {
              return items?.length > 1 && isLocationSpacesActive(router, 'files-spaces-projects')
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.projects.actions',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'space-actions',
            icon: 'slideshow-3',
            title: () => $gettext('Actions'),
            component: SpaceActions,
            isEnabled: ({ items }) => {
              if (items?.length !== 1) {
                return false
              }
              if (!isProjectSpaceResource(items[0])) {
                return false
              }
              if (
                !isLocationSpacesActive(router, 'files-spaces-projects') &&
                !isLocationSpacesActive(router, 'files-spaces-generic')
              ) {
                return false
              }
              return [
                ...items[0].spaceRoles[spaceRoleManager.name],
                ...items[0].spaceRoles[spaceRoleEditor.name]
              ].some((role) => role.id === unref(user).uuid)
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.projects.sharing',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'space-share',
            icon: 'group',
            title: () => $gettext('Members'),
            component: SharesPanel,
            componentAttrs: () => ({
              showSpaceMembers: true,
              get showLinks() {
                if (unref(isSharingEnabled)) {
                  return unref(arePublicLinksEnabled)
                }
                return false
              }
            }),
            isEnabled: ({ items }) => {
              return items?.length === 1 && isProjectSpaceResource(items[0])
            }
          }
        }
      ] satisfies SidebarPanelExtension<Resource, Resource>[]
  )
}
