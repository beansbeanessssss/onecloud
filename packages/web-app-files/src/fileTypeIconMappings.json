{
  "7z": "file-zip",
  "accdb": "file",
  "ai": "image",
  "apk": "file-zip",
  "bash": "file-code",
  "bz2": "file-zip",
  "cb7": "file-text",
  "cba": "file-text",
  "cbr": "file-text",
  "cbt": "file-text",
  "cbtc": "file-text",
  "cbz": "file-text",
  "cc": "file-code",
  "cdr": "image",
  "cpp": "file-code",
  "css": "file-code",
  "csv": "file-excel-2",
  "cvbdl": "file-text",
  "c": "file-code",
  "c++": "file-code",
  "deb": "file-zip",
  "doc": "file-word-2",
  "docm": "file-word-2",
  "docx": "file-word-2",
  "dot": "file-word-2",
  "dotx": "file-word-2",
  "eml": "file-text",
  "eot": "image",
  "epub": "file-text",
  "eps": "image",
  "gz": "file-zip",
  "gzip": "file-zip",
  "h": "file-code",
  "hh": "file-code",
  "hpp": "file-code",
  "html": "file-code",
  "htm": "file-code",
  "java": "file-code",
  "js": "file-code",
  "json": "file-code",
  "jpeg": "image",
  "jpg": "image",
  "lwp": "file-word-2",
  "mdb": "file",
  "odp": "file-ppt-2",
  "ods": "file-excel-2",
  "odt": "file-word-2",
  "one": "file-word-2",
  "otf": "image",
  "pdf": "file-pdf",
  "pfb": "image",
  "php": "file-code",
  "pl": "file-code",
  "png": "image",
  "pot": "file-ppt-2",
  "potm": "file-ppt-2",
  "potx": "file-ppt-2",
  "ppa": "file-ppt-2",
  "ppam": "file-ppt-2",
  "pps": "file-ppt-2",
  "ppsm": "file-ppt-2",
  "ppsx": "file-ppt-2",
  "ppt": "file-ppt-2",
  "pptm": "file-ppt-2",
  "pptx": "file-ppt-2",
  "ps": "image",
  "psd": "image",
  "py": "file-code",
  "rar": "file-zip",
  "rss": "file",
  "sh-lib": "file-code",
  "sh": "file-code",
  "swf": "file",
  "tar": "file-zip",
  "tar.bz2": "file-zip",
  "tar.gz": "file-zip",
  "tbz2": "file-zip",
  "tex": "file-text",
  "tgz": "file-zip",
  "ttf": "image",
  "vsd": "file-word-2",
  "woff": "image",
  "wpd": "file-word-2",
  "xcf": "image",
  "xla": "file-excel-2",
  "xlam": "file-excel-2",
  "xls": "file-excel-2",
  "xlsb": "file-excel-2",
  "xlsm": "file-excel-2",
  "xlsx": "file-excel-2",
  "xlt": "file-excel-2",
  "xltm": "file-excel-2",
  "xltx": "file-excel-2",
  "xml": "file-code",
  "yaml": "file-code",
  "yml": "file-code",
  "zip": "file-zip"
}
