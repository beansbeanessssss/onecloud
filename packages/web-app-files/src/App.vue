<template>
  <main id="files" class="oc-flex oc-height-1-1">
    <div v-if="dragareaEnabled" class="dragarea" />
    <router-view tabindex="0" class="oc-width-expand" />
  </main>
</template>
<script lang="ts">
import { defineComponent, watch } from '@vue/composition-api'
import { Route } from 'vue-router'
import { useRoute, useStore } from 'web-pkg/src/composables'
import { bus } from 'web-pkg/src/instance'

export default defineComponent({
  setup() {
    const store = useStore<any>()
    watch(useRoute(), (to: Route, from?: Route) => {
      store.dispatch('Files/resetFileSelection')
    })
  },
  data: () => ({
    dragareaEnabled: false
  }),
  created() {
    const dragOver = bus.subscribe('drag-over', this.onDragOver)
    const dragOut = bus.subscribe('drag-out', this.hideDropzone)
    const drop = bus.subscribe('drop', this.hideDropzone)

    this.$on('beforeDestroy', () => {
      bus.unsubscribe('drag-over', dragOver)
      bus.unsubscribe('drag-out', dragOut)
      bus.unsubscribe('drop', drop)
    })
  },

  methods: {
    hideDropzone() {
      this.dragareaEnabled = false
    },
    onDragOver(event) {
      this.dragareaEnabled = (event.dataTransfer.types || []).some((e) => e === 'Files')
    }
  }
})
</script>

<style lang="scss" scoped>
main {
  max-height: 100%;
}

.dragarea {
  background-color: rgba(60, 130, 225, 0.21);
  pointer-events: none;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  position: absolute;
  z-index: 9;
  border-radius: 14px;
  border: 2px dashed var(--oc-color-swatch-primary-muted);
}

#files {
  position: relative;
}
</style>
