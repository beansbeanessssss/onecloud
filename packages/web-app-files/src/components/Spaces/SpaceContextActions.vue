<template>
  <div>
    <context-action-menu :menu-sections="menuSections" :items="items" :space="space" />
    <quota-modal
      v-if="quotaModalIsOpen"
      :cancel="closeQuotaModal"
      :space="quotaModalSelectedSpace"
    />
    <readme-content-modal
      v-if="readmeContentModalIsOpen"
      :cancel="closeReadmeContentModal"
      :space="items[0]"
    />
    <input
      id="space-image-upload-input"
      ref="spaceImageInput"
      type="file"
      name="file"
      multiple
      tabindex="-1"
      accept="image/*"
      @change="$_uploadImage_uploadImageSpace"
    />
  </div>
</template>

<script lang="ts">
import ContextActionMenu from 'web-pkg/src/components/ContextActions/ContextActionMenu.vue'
import QuotaModal from 'web-pkg/src/components/Spaces/QuotaModal.vue'
import ReadmeContentModal from 'web-pkg/src/components/Spaces/ReadmeContentModal.vue'

import Delete from 'web-pkg/src/mixins/spaces/delete'
import Rename from 'web-pkg/src/mixins/spaces/rename'
import Restore from 'web-pkg/src/mixins/spaces/restore'
import ShowDetails from '../../mixins/actions/showDetails'
import EditDescription from 'web-pkg/src/mixins/spaces/editDescription'
import EditQuota from 'web-pkg/src/mixins/spaces/editQuota'
import DeletedFiles from 'web-pkg/src/mixins/spaces/deletedFiles'
import Disable from 'web-pkg/src/mixins/spaces/disable'
import ShowMembers from 'web-pkg/src/mixins/spaces/showMembers'
import UploadImage from '../../mixins/spaces/actions/uploadImage'
import EditReadmeContent from 'web-pkg/src/mixins/spaces/editReadmeContent'
import { isLocationSpacesActive } from '../../router'
import { defineComponent, PropType } from 'vue'
import { SpaceResource } from 'web-client/src/helpers'

export default defineComponent({
  name: 'SpaceContextActions',
  components: { ContextActionMenu, QuotaModal, ReadmeContentModal },
  mixins: [
    Rename,
    Delete,
    EditDescription,
    EditQuota,
    DeletedFiles,
    Disable,
    ShowDetails,
    ShowMembers,
    Restore,
    UploadImage,
    EditReadmeContent
  ],

  props: {
    space: {
      type: Object as PropType<SpaceResource>,
      required: true
    },
    items: {
      type: Array,
      required: true
    }
  },

  computed: {
    quotaModalSelectedSpace() {
      return this.$data.$_editQuota_selectedSpace
    },
    quotaModalIsOpen() {
      return this.$data.$_editQuota_modalOpen
    },
    readmeContentModalIsOpen() {
      return this.$data.$_editReadmeContent_modalOpen
    },
    menuSections() {
      const sections = []

      if (this.menuItemsMembers.length) {
        sections.push({
          name: 'members',
          items: this.menuItemsMembers
        })
      }
      if (this.menuItemsPrimaryActions.length) {
        sections.push({
          name: 'primaryActions',
          items: this.menuItemsPrimaryActions
        })
      }
      if (this.menuItemsSecondaryActions.length) {
        sections.push({
          name: 'secondaryActions',
          items: this.menuItemsSecondaryActions
        })
      }
      if (this.menuItemsTrashBin.length) {
        sections.push({
          name: 'trashBin',
          items: this.menuItemsTrashBin
        })
      }
      if (this.menuItemsSidebar.length) {
        sections.push({
          name: 'sidebar',
          items: this.menuItemsSidebar
        })
      }

      return sections
    },

    filterParams() {
      return {
        resources: this.items
      }
    },

    menuItemsMembers() {
      const fileHandlers = [...this.$_showMembers_items]
      return [...fileHandlers].filter((item) => item.isEnabled(this.filterParams))
    },
    menuItemsPrimaryActions() {
      const fileHandlers = [
        ...this.$_rename_items,
        ...this.$_editDescription_items,
        ...this.$_uploadImage_items
      ]

      if (isLocationSpacesActive(this.$router, 'files-spaces-generic')) {
        fileHandlers.splice(2, 0, ...this.$_editReadmeContent_items)
      }
      return [...fileHandlers].filter((item) => item.isEnabled(this.filterParams))
    },
    menuItemsSecondaryActions() {
      const fileHandlers = [
        ...this.$_editQuota_items,
        ...this.$_disable_items,
        ...this.$_restore_items,
        ...this.$_delete_items
      ]

      return [...fileHandlers].filter((item) => item.isEnabled(this.filterParams))
    },
    menuItemsTrashBin() {
      const fileHandlers = [...this.$_deletedFiles_items]
      return [...fileHandlers].filter((item) => item.isEnabled(this.filterParams))
    },
    menuItemsSidebar() {
      const fileHandlers = [...this.$_showDetails_items]
      return [...fileHandlers].filter((item) => item.isEnabled(this.filterParams))
    }
  },
  methods: {
    closeQuotaModal() {
      this.$_editQuota_closeModal()
    },
    closeReadmeContentModal() {
      this.$_editReadmeContent_closeModal()
    }
  }
})
</script>

<style lang="scss">
#space-image-upload-input {
  position: absolute;
  left: -99999px;
}
</style>
