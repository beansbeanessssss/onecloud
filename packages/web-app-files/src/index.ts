import App from './App.vue'
import Favorites from './views/Favorites.vue'
import FilesDrop from './views/FilesDrop.vue'
import SharedWithMe from './views/shares/SharedWithMe.vue'
import SharedWithOthers from './views/shares/SharedWithOthers.vue'
import SharedViaLink from './views/shares/SharedViaLink.vue'
import SpaceDriveResolver from './views/spaces/DriveResolver.vue'
import SpaceProjects from './views/spaces/Projects.vue'
import TrashOverview from './views/trash/Overview.vue'
import translations from '../l10n/translations.json'
import { defineWebApplication, quickActions } from '@ownclouders/web-pkg'
import store from './store'
import { extensions } from './extensions'
import fileSideBars from './fileSideBars'
import { buildRoutes } from '@ownclouders/web-pkg'
import { AppNavigationItem } from '@ownclouders/web-pkg'

// dirty: importing view from other extension within project
import SearchResults from '../../web-app-search/src/views/List.vue'
import {
  SpaceResource,
  isPersonalSpaceResource,
  isShareSpaceResource
} from '@ownclouders/web-client/src/helpers'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: $gettext('Files'),
  id: 'files',
  icon: 'resource-type-folder',
  color: 'var(--oc-color-swatch-primary-muted)',
  isFileEditor: false,
  extensions: [],
  fileSideBars
}
export const navItems = (context): AppNavigationItem[] => {
  return [
    {
      name(capabilities) {
        return capabilities.spaces?.enabled ? $gettext('Personal') : $gettext('All files')
      },
      icon: appInfo.icon,
      route: {
        path: `/${appInfo.id}/spaces/personal`
      },
      isActive: () => {
        return (
          !context.$store.getters['runtime/spaces/currentSpace'] ||
          context.$store.getters['runtime/spaces/currentSpace']?.isOwner(
            context.$store.getters.user
          )
        )
      },
      enabled(capabilities) {
        if (!capabilities.spaces?.enabled) {
          return true
        }
        return !!context?.$store?.getters['runtime/spaces/spaces'].find(
          (drive) => isPersonalSpaceResource(drive) && drive.isOwner(context.$store.getters.user)
        )
      }
    },
    {
      name: $gettext('Favorites'),
      icon: 'star',
      route: {
        path: `/${appInfo.id}/favorites`
      },
      enabled(capabilities) {
        return capabilities.files?.favorites
      }
    },
    {
      name: $gettext('Shares'),
      icon: 'share-forward',
      route: {
        path: `/${appInfo.id}/shares`
      },
      isActive: () => {
        const space = context.$store.getters['runtime/spaces/currentSpace'] as SpaceResource
        // last check is when fullShareOwnerPaths is enabled
        return !space || isShareSpaceResource(space) || !space?.isOwner(context.$store.getters.user)
      },
      activeFor: [
        { path: `/${appInfo.id}/spaces/share` },
        { path: `/${appInfo.id}/spaces/personal` }
      ],
      enabled(capabilities) {
        return capabilities.files_sharing?.api_enabled !== false
      }
    },
    {
      name: $gettext('Spaces'),
      icon: 'layout-grid',
      route: {
        path: `/${appInfo.id}/spaces/projects`
      },
      activeFor: [{ path: `/${appInfo.id}/spaces/project` }],
      enabled(capabilities) {
        return capabilities.spaces?.projects
      }
    },
    {
      name: $gettext('Deleted files'),
      icon: 'delete-bin-5',
      route: {
        path: `/${appInfo.id}/trash/overview`
      },
      activeFor: [{ path: `/${appInfo.id}/trash` }],
      enabled(capabilities) {
        return capabilities.dav?.trashbin === '1.0' && capabilities.files?.undelete
      }
    }
  ]
}

export default defineWebApplication({
  setup(args) {
    return {
      appInfo,
      store,
      routes: buildRoutes({
        App,
        Favorites,
        FilesDrop,
        SearchResults,
        Shares: {
          SharedViaLink,
          SharedWithMe,
          SharedWithOthers
        },
        Spaces: {
          DriveResolver: SpaceDriveResolver,
          Projects: SpaceProjects
        },
        Trash: {
          Overview: TrashOverview
        }
      }),
      navItems,
      quickActions,
      translations,
      extensions: extensions(args)
    }
  }
})
