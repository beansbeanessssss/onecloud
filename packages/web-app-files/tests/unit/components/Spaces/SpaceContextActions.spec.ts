import SpaceContextActions from '../../../../src/components/Spaces/SpaceContextActions.vue'
import { buildSpace } from '@ownclouders/web-client/src/helpers'
import {
  createStore,
  defaultComponentMocks,
  defaultPlugins,
  mount,
  defaultStoreMockOptions,
  RouteLocation
} from 'web-test-helpers'
import { mock } from 'jest-mock-extended'
import { Drive } from '@ownclouders/web-client/src/generated'

const spaceMock = mock<Drive>({
  id: '1',
  root: { permissions: [{ roles: ['manager'], grantedToIdentities: [{ user: { id: '1' } }] }] },
  driveType: 'project',
  special: null
})

describe('SpaceContextActions', () => {
  describe('action handlers', () => {
    it('renders actions that are always available: "Members", "Edit Quota", "Details"', () => {
      const { wrapper } = getWrapper(buildSpace(spaceMock))
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})

function getWrapper(space) {
  const mocks = defaultComponentMocks({ currentRoute: mock<RouteLocation>({ path: '/files' }) })
  mocks.$previewService.getSupportedMimeTypes.mockReturnValue([])
  const store = createStore(defaultStoreMockOptions)
  return {
    wrapper: mount(SpaceContextActions, {
      props: {
        actionOptions: {
          resources: [space]
        }
      },
      global: {
        mocks,
        provide: mocks,
        plugins: [
          ...defaultPlugins({
            abilities: [{ action: 'set-quota-all', subject: 'Drive' }]
          }),
          store
        ]
      }
    })
  }
}
