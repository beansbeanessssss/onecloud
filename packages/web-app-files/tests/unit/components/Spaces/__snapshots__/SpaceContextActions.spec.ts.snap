// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpaceContextActions action handlers renders actions that are always available: "Members", "Deleted files", "Details" 1`] = `
<div>
  <div id="oc-files-context-menu">
    <ul class="oc-list oc-my-rm oc-mx-rm oc-files-context-actions oc-pb-s oc-files-context-actions-border" id="oc-files-context-actions-members">
      <li class="oc-files-context-action oc-px-s oc-rounded oc-menu-item-hover">
        <button class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-files-actions-show-details-trigger action-menu-item" type="button">
          <!-- @slot Content of the button -->
          <span class="oc-icon oc-icon-m oc-icon-passive">
            <!---->
          </span>
          <span class="oc-files-context-action-label">Members</span>
          <!--v-if-->
          <!--v-if-->
        </button>
      </li>
    </ul>
    <ul class="oc-list oc-my-rm oc-mx-rm oc-files-context-actions oc-pb-s oc-pt-s oc-files-context-actions-border" id="oc-files-context-actions-trashBin">
      <li class="oc-files-context-action oc-px-s oc-rounded oc-menu-item-hover">
        <button class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-files-actions-delete-trigger action-menu-item" type="button">
          <!-- @slot Content of the button -->
          <span class="oc-icon oc-icon-m oc-icon-passive">
            <!---->
          </span>
          <span class="oc-files-context-action-label">Deleted files</span>
          <!--v-if-->
          <!--v-if-->
        </button>
      </li>
    </ul>
    <ul class="oc-list oc-my-rm oc-mx-rm oc-files-context-actions oc-pt-s" id="oc-files-context-actions-sidebar">
      <li class="oc-files-context-action oc-px-s oc-rounded oc-menu-item-hover">
        <button class="oc-button oc-rounded oc-button-s oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-files-actions-show-details-trigger action-menu-item" type="button">
          <!-- @slot Content of the button -->
          <span class="oc-icon oc-icon-m oc-icon-passive">
            <!---->
          </span>
          <span class="oc-files-context-action-label">Details</span>
          <!--v-if-->
          <!--v-if-->
        </button>
      </li>
    </ul>
  </div>
  <!--v-if-->
  <!--v-if-->
  <input accept="image/*" id="space-image-upload-input" multiple="" name="file" tabindex="-1" type="file">
</div>
`;
